File: COLLEGE/ASSIGNMENT7/c/Q.9.c
#include <stdio.h>
#include <stdlib.h>

struct Queue {
    int* arr;
    int front, rear, size, capacity;
};

struct Queue* createQueue(int capacity) {
    struct Queue* queue = (struct Queue*)malloc(sizeof(struct Queue));
    queue->capacity = capacity;
    queue->front = queue->size = 0;
    queue->rear = capacity - 1;
    queue->arr = (int*)malloc(queue->capacity * sizeof(int));
    return queue;
}

int isFull(struct Queue* queue) {
    return (queue->size == queue->capacity);
}

int isEmpty(struct Queue* queue) {
    return (queue->size == 0);
}

void enqueue(struct Queue* queue, int item) {
    if (isFull(queue)) return;
    queue->rear = (queue->rear + 1) % queue->capacity;
    queue->arr[queue->rear] = item;
    queue->size++;
}

int dequeue(struct Queue* queue) {
    if (isEmpty(queue)) return -1;
    int item = queue->arr[queue->front];
    queue->front = (queue->front + 1) % queue->capacity;
    queue->size--;
    return item;
}

void reverseQueue(struct Queue* queue) {
    if (!isEmpty(queue)) {
        int item = dequeue(queue);
        reverseQueue(queue);
        enqueue(queue, item);
    }
}

void display(struct Queue* queue) {
    if (isEmpty(queue)) {
        printf("Queue is empty\n");
        return;
    }
    printf("Queue elements: ");
    for (int i = queue->front; i < queue->size + queue->front; i++) {
        printf("%d ", queue->arr[i % queue->capacity]);
    }
    printf("\n");
}

int main() {
    struct Queue* queue = createQueue(5);
    enqueue(queue, 10);
    enqueue(queue, 20);
    enqueue(queue, 30);
    printf("Original Queue: ");
    display(queue);
    reverseQueue(queue);
    printf("Reversed Queue: ");
    display(queue);
    return 0;
}

File: COLLEGE/ASSIGNMENT7/python/Q.9.py
from collections import deque

class Queue:
    def __init__(self):
        self.queue = deque()  # Initialize a queue using deque

    # Function to enqueue an element to the queue
    def enqueue(self, item):
        self.queue.append(item)
        print(f"Enqueued {item} to the queue.")

    # Function to dequeue an element from the queue
    def dequeue(self):
        if self.is_empty():
            print("Queue is empty! Cannot dequeue element.")
            return None
        return self.queue.popleft()  # Pop from the front of the queue

    # Function to check if the queue is empty
    def is_empty(self):
        return len(self.queue) == 0

    # Function to reverse the queue
    def reverse(self):
        if self.is_empty():
            return
        # Remove the front element
        front = self.dequeue()
        # Reverse the remaining queue
        self.reverse()
        # Add the removed element to the back of the queue
        self.enqueue(front)

    # Function to display the current elements in the queue
    def display(self):
        if self.is_empty():
            print("Queue is empty!")
        else:
            print("Queue elements:", list(self.queue))  # Convert deque to list for display


# Test the Queue class
queue = Queue()

# Enqueue elements to the queue
queue.enqueue(1)
queue.enqueue(2)
queue.enqueue(3)
queue.enqueue(4)
queue.enqueue(5)

# Display the original queue
print("Original queue:")
queue.display()

# Reverse the queue
queue.reverse()

# Display the reversed queue
print("Reversed queue:")
queue.display()
